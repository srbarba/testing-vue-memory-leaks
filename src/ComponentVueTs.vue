<template>
  <v-card>
    {{ text }}
    <div>{{ $t('Hello world') }}</div>
  </v-card>
</template>

<script lang="ts">
import Vue from 'vue'
export default Vue.extend({
  props: {
    text: {
      type: String as Vue.PropType<string>,
      default: 'Dummy component',
    },
  },
})
</script>
