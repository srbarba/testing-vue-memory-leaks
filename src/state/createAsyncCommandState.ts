import Vue from 'vue'
import deepmerge from 'deepmerge'

export const STATUS = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
}

const initialState = <U>(initialData: U) =>
  Vue.observable({
    status: STATUS.IDLE,
    data: deepmerge<U>({}, initialData),
    error: undefined as Error | undefined,
  })

export function createAsyncCommandState<U>(initialData = {} as U) {
  const state = initialState(initialData)

  return {
    getState() {
      return state
    },
    getData() {
      return state.data
    },
    getError() {
      return state.error
    },
    getStatus() {
      return state.status
    },
    isLoading() {
      return state.status === STATUS.LOADING
    },
    isError() {
      return state.status === STATUS.ERROR
    },
    isSuccess() {
      return state.status === STATUS.SUCCESS
    },
    setLoading() {
      state.status = STATUS.LOADING
    },
    setSuccess() {
      state.status = STATUS.SUCCESS
    },
    setData<U>(data: U) {
      state.status = STATUS.IDLE
      state.error = undefined
      Vue.set(state, 'data', Object.assign(state.data, data))
    },
    setError(appError?: Error) {
      state.status = STATUS.ERROR
      state.error = appError
    },
    reset() {
      Object.assign(state, initialState(initialData))
    },
  }
}
