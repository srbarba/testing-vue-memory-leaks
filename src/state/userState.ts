import type { User } from '../types/User'
import { createAsyncCommandState } from './createAsyncCommandState'

const initialState = {
  id: 0,
  name: '',
} as User

const commandState = createAsyncCommandState(initialState)
export const reset = commandState.reset

export const computedUserName = {
  set(name: User['name']) {
    commandState.setData({ name })
  },
  get() {
    return commandState.getData().name
  },
}

export const computedId = {
  set(id: User['id']) {
    commandState.setData({ id })
  },
  get() {
    return commandState.getData().id
  },
}
