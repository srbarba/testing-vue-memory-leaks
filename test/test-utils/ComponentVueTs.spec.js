import { mountWithVuetify } from '../mountWithVuetify'
import ComponentVueTs from '../../src/ComponentVueTs.vue'
import * as userState from '../../src/state/userState'
import { UserMother } from '../UserMotherWithoutFaker'

describe('ComponentVueTs', () => {
  beforeEach(() => {
    userState.reset()
  })

  it('renders correctly', () => {
    const wrapper = mountWithVuetify(ComponentVueTs)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders correctly with props', () => {
    const wrapper = mountWithVuetify(ComponentVueTs, {
      propsData: {
        text: 'My dummy Component',
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('given user name', () => {
    beforeEach(() => {
      const user = UserMother.create()
      userState.computedUserName.set(user.name)
    })

    it('renders correctly', () => {
      const wrapper = mountWithVuetify(ComponentVueTs)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders correctly with props', () => {
      const wrapper = mountWithVuetify(ComponentVueTs, {
        propsData: {
          text: 'My dummy Component',
        },
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
