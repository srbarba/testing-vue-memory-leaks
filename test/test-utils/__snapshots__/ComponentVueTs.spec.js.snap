// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentVueTs renders correctly 1`] = `
<div class="v-card v-sheet theme--light">
  Dummy component
  <div>Hello world</div>
</div>
`;

exports[`ComponentVueTs renders correctly with props 1`] = `
<div class="v-card v-sheet theme--light">
  My dummy Component
  <div>Hello world</div>
</div>
`;
